### DISCUSSION TITLE
is there a way to extract data from detected objects in test data and store in a text file?

### DISCUSSION THREAD

**Discussion by SWRIGH211** [ultralytics/yolov5#2506] [Q&A]
![image](https://user-images.githubusercontent.com/60390273/111506611-b2f5a880-8741-11eb-8ac9-f2949bfb75ff.png)

the image is one of my test data images displaying detected traffic signs
I was wondering if there is an automated way to extract objects detected in test data and store them in a text file. to be more specific not the co-ords of the bounding boxes, but the accuracy of the object and the category that it has been identified to belong to. i.e.
-----------txt-file-------------------------
other 70%
other 88%
mandatory 56%

Any help would be much appreciated!

**Comment by SWRIGH211** [comment]
![image](https://user-images.githubusercontent.com/60390273/111621950-817fea00-87e0-11eb-99c6-bfda7df46b15.png)
ive managed to display the sign in text format but not the accuracy after following one of glens comments on another issue: https://github.com/ultralytics/yolov5/issues/36.
is displaying the accuracy possible too?

**Reply by SWRIGH211** [reply]
Because basically I'm wanting to save true positive values into a pandas data frame to determine the test accuracy. Would my best bet be to add --iou 0.5 to the "model =" line? Or is that wrong

**Reply by glenn-jocher** [reply]
@SWRIGH211 yes the default results.print() output has been enchanced a bit now, but this is really just there for your convenience, it's not meant to be a 100% solution for everyone's needs. 

For custom requirements you can examine the results values directly and apply your own custom code to them, i.e. results.xyxy will show you the actual tensor with the boxes, classes and confidences for each image in your batch:
```python
# Inference
results = model(imgs, size=640)  # includes NMS

# Results
results.print()  
results.save()  # or .show()

# Data
print(results.xyxy[0])  # print img1 predictions (pixels)
#                   x1           y1           x2           y2   confidence        class
# tensor([[7.50637e+02, 4.37279e+01, 1.15887e+03, 7.08682e+02, 8.18137e-01, 0.00000e+00],
#         [9.33597e+01, 2.07387e+02, 1.04737e+03, 7.10224e+02, 5.78011e-01, 0.00000e+00],
#         [4.24503e+02, 4.29092e+02, 5.16300e+02, 7.16425e+02, 5.68713e-01, 2.70000e+01]])
```

Please see YOLOv5 PyTorch Hub Tutorial for details:

## YOLOv5 Tutorials

* [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)&nbsp; 🚀 RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://github.com/ultralytics/yolov5/issues/475)
* [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://github.com/ultralytics/yolov5/issues/251)
* [Test-Time Augmentation (TTA)](https://github.com/ultralytics/yolov5/issues/303)
* [Model Ensembling](https://github.com/ultralytics/yolov5/issues/318)
* [Model Pruning/Sparsity](https://github.com/ultralytics/yolov5/issues/304)
* [Hyperparameter Evolution](https://github.com/ultralytics/yolov5/issues/607)
* [Transfer Learning with Frozen Layers](https://github.com/ultralytics/yolov5/issues/1314)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Reply by SWRIGH211** [reply]
perfect, thanks glenn

**Reply by SWRIGH211** [reply]
@glenn-jocher 
is there a way to call all in all images into this command without having to add them in to the highlighted sections one by one shown in the image. also the confidence value that is highlighted at the bottom of the snippet is lower than what is displayed on the image. I.e. yield in the image is at 0.96 whereas the conf states 9.2e-01 ans 7.64e-01 which correct me if im wrong is 0.92 and 0.764% but both in the image say 96%? Any ideas
![image](https://user-images.githubusercontent.com/60390273/111880538-0a934e80-89a4-11eb-83ae-cf29fe82c84c.png)
![image](https://user-images.githubusercontent.com/60390273/111880574-37476600-89a4-11eb-922a-b890495fb4c7.png)

**Reply by SWRIGH211** [reply]
![image](https://user-images.githubusercontent.com/60390273/111882191-30bcec80-89ac-11eb-85d3-b218e88b40bb.png)
also when ran through certain frames it doesnt pick up labels that are clearly there when you check the image yourself? Any ideas why. its just becasue im trying to calculate my test accuracy scores per each catergory and i have a lot of images and i dont want to manually have to write out and calulate the True prediction false prediction score. many thanks in advance

**Reply by glenn-jocher** [reply]
@SWRIGH211 if inference is missing some objects you can reduce confidence threshold towards 0.0 to increase recall:
```python
model.conf = 0.1  # default = 0.25
results = model(imgs)
```

I will try to reproduce the confidence disparity you mentioned.

**Reply by glenn-jocher** [reply]
@SWRIGH211 I am **not** able to reproduce any issues with displayed confidences. All confidences between results object and displayed images match correctly.

<img width="1500" alt="Screenshot 2021-03-23 at 13 35 30" src="https://user-images.githubusercontent.com/26833433/112147291-d22a8500-8bdc-11eb-9712-ee4f77558b22.png">
